### ISSUE TITLE
ASTF too many ACKs with TSO

### ISSUE THREAD

**Issue by jsmoon** [cisco-system-traffic-generator/trex-core#962]
Hi, I found an issue that TSO does not work as I expected.
I've tested in my loopback environment by `start -f astf/ftp_multiple_flows_sim.py -t fsize=10000`.
```
Cpu Util(%)

  Thread   | Avg | Latest | -1  | -2  | -3  | -4  | -5  | -6  | -7  | -8  | -9  | -10 | -11 | -12 | -13 | -14
-----------+-----+--------+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+-----+----
 0 (IDLE)  |   4 |      0 |   0 |   0 |   0 |  20 | 100 | 100 | 100 | 100 | 100 |  97 |   0 |   0 |   0 |   0

Traffic stats of Profile ID : _. Number of template groups = 0

                           |      client       |      server       |
---------------------------+-------------------+-------------------+---------------------------
            m_active_flows |                 0 |                 0 | active open flows
               m_est_flows |                 0 |                 0 | active established flows
              m_tx_bw_l7_r |             0 bps |       469.93 Mbps | tx L7 bw acked
        m_tx_bw_l7_total_r |             0 bps |       469.69 Mbps | tx L7 bw total
              m_rx_bw_l7_r |       469.86 Mbps |             0 bps | rx L7 bw acked
                m_tx_pps_r |        40.57 Kpps |        40.53 Kpps | tx pps
                m_rx_pps_r |        40.61 Kpps |        40.58 Kpps | rx pps
                m_avg_size |          723.46 B |          724.21 B | average pkt size
                m_tx_ratio |               0 % |          100.05 % | Tx acked/sent ratio
                         - |                   |                   |
        m_traffic_duration |          7.12 sec |          6.62 sec | measured traffic duration
                         - |                   |                   |
                       TCP |                   |                   |
                         - |                   |                   |
          tcps_connattempt |                 1 |                 0 | connections initiated
              tcps_accepts |                 0 |                 1 | connections accepted
             tcps_connects |                 1 |                 1 | connections established
               tcps_closed |                 1 |                 1 | conn. closed (includes drops)
            tcps_segstimed |                 3 |           7222290 | segs where we tried to get rtt
           tcps_rttupdated |                 3 |           7227987 | times we succeeded
             tcps_sndtotal |           7227988 |           7222444 | total packets sent
              tcps_sndpack |                 1 |           7222441 | data packets sent
              tcps_sndbyte |                 9 |       10485760000 | data bytes sent by application
           tcps_sndbyte_ok |                 9 |       10485760000 | data bytes sent by tcp
              tcps_sndctrl |                 1 |                 1 | control (SYN|FIN|RST) packets sent
              tcps_sndacks |           7227986 |                 2 | ack-only packets sent
             tcps_rcvtotal |           7233102 |           7227988 | total packets received
              tcps_rcvpack |           7249969 |                 2 | packets received in sequence
              tcps_rcvbyte |       10485760000 |                 9 | bytes received in sequence
           tcps_rcvackpack |                19 |           7227987 | rcvd ack packets
           tcps_rcvackbyte |                 9 |       10485760000 | tx bytes acked by rcvd acks
        tcps_rcvackbyte_of |                 1 |                 1 | tx bytes acked by rcvd acks - overflow acked
          tcps_rcvoffloads |             16885 |                 0 | receive offload packets by software
              tcps_preddat |           7233098 |                 0 | times hdr predict ok for data pkts
              tcps_predack |                 0 |           7227984 | times hdr predict ok for acks
                         - |                   |                   |
                       UDP |                   |                   |
                         - |                   |                   |
                         - |                   |                   |
               Application |                   |                   |
                         - |                   |                   |
                         - |                   |                   |
                Flow Table |                   |                   |
                         - |                   |                   |
          err_rx_throttled |             22616 |             22583 | rx thread was throttled
```
1. The average number of bytes per packet (tcps_sndbyte/tcps_sndpack) sent by the server is 1451. It means most packets were not sent with TSO.
2. The client sent ACK packets for almost all received packets: 7227986 of 7249969

I captured the packets and I found that lots of data packets have the PUSH flag.
According to my investigation, I guess that the following scenario can make this issue.
1. When a packet will be sent by TSO and no data is left in the socket send buffer, the PUSH flag can be set.
2. In this case, all the TSO segmented packets have the PUSH flag.
3. When the packet with the PUSH flag is received, the related ACK packet will be sent immediately. (ACKed bytes will be segment size, 1448)
4. When the ACK packet is received, the socket send buffer will drop the ACKed bytes and append data from the application buffer if available. (Since the FTP profile uses nonblocking send, probably this can happen.)
5. When the data is appended by ACK, it will be sent immediately. The data size will be the same as ACKed bytes and have the PUSH flag also.

When I use `--tso-disable` with `--lro-disable`, I can see half of ACK packets as expected by `no_delay_counter = mss*2`.
```
              tcps_sndacks |           3591016 |                 2 | ack-only packets sent         
             tcps_rcvtotal |           7241550 |           3591018 | total packets received    
              tcps_rcvpack |           7241548 |                 2 | packets received in sequence 
              tcps_rcvbyte |       10485760000 |                 9 | bytes received in sequence 
```
In this case, the performance is even better.

@hhaim I'd like to get your opinion on this issue.

**Comment by jsmoon** [comment]
I think the following improvement points:
1. Send the TSO packet without the PUSH flag.
    * When the TSO packet length is aligned by TSO segment size, in general, it may not be the last data in the socket send buffer and does not have the PUSH flag.
    * With this change, in most cases, the TSO packet will not have the PUSH flag and SLRO will merge more packets.
    * In addition, I think this change could make the GTPU tunnel checksum valid even TSO packets.
2. Don't reply ACK packet immediately to the PUSH packet.
    * The immediate ACK is the T-Rex compatible feature which is not supported by the FreeBSD TCP stack.
    * Since the original FreeBSD TCP stack sends an ACK packet on every other packet received, I've met some failures in test cases and I've decided to make it T-Rex compatible.
3. Send data packets after all received ACK packets handled.
    * This change will not reduce ACK packets, but increase TSO packets.

@hhaim what do you think of these suggestions?

**Comment by hhaim** [comment]
@jsmoon you are right. TSO does not work. it worked in the past and the 32K bytes HTTP was TSO for the almost all the 32K with one TSO ( initwnd =10) 

regarding (2), delaying  the ACK to second packet is a feature. I wouldn't send a timer to very low timeout as it influences the maximum scale

**Comment by jsmoon** [comment]
@hhaim I'm sorry I can't understand what is your opinion for. It will be more help if you answer to my questions respectively.

Regarding 1, do you agree to this change? As a new feature, I think I can make it controlled by a command line option, `--tso-align-size`.

Regarding 2, it (=sending an ACK packet on every other packet received) is a default action of the FreeBSD 13.0 TCP stack. I removed it to keep the T-Rex `tcp.no_delay_counter` effective. Do you want to keep it as today?

Regarding 3, I think this change maximizes the TSO usage in some cases. The SLRO feature can handle it simply. What do you think of it?

**Comment by hhaim** [comment]
@jsmoon 
1.  I think your analysis is wrong, 
comparing the old and the current code, I see the issue. TSO logic does **not** work at all in latest
This is the part that is missing .. 
```
    if (!tso) {
        if (len > tp->t_maxseg - optlen) {
            len = tp->t_maxseg - optlen;
            sendalot = 1;
            flags &= ~TH_FIN;
        }
    }else {
        if (len > tp->t_maxseg - optlen ){
            if (optlen) {
                if ((len%tp->t_maxseg)==0) {
                    len -= (len/tp->t_maxseg)*optlen; /* reduce the optlen from all packets*/
                    sendalot = 1;
                    flags &= ~TH_FIN;
                }
            }
        }
    }

```

Are you aware of it?  

regarding your questio.  ```TSO packet length```  is it ```dlen```?  let's take an example in case ```dlen==100000``` and ```mss=1500``` I would send 66 windows without PUSH and residue with PUSH (the last). 

I don't understand why do you need another tunable ```tso-align-size```?
2. I would keep the old logic. ( I understand that the default freebsd send ACK and any other packet but it is possible to get to this using current tunbables)
3. let's start by fixing 1

**Comment by jsmoon** [comment]
@hhaim 
1. I didn't compare with old code. I just checked the current code which comes from FreeBSD TCP stack. So, I suggested this as a new feature, `--tso-align-size`. Now, let's forget it.
    For my analysis, I found the following codes: https://github.com/cisco-system-traffic-generator/trex-core/blob/c0288e1dccf5339ecac6a54d09bfd00ac600a3ad/src/44bsd/netinet/tcp_output.c#L661-L669 When the send sockbuf is emptied, `len` may not be aligned with the `mss` and TH_PUSH is added at next line.
    https://github.com/cisco-system-traffic-generator/trex-core/blob/c0288e1dccf5339ecac6a54d09bfd00ac600a3ad/src/44bsd/netinet/tcp_output.c#L754-L757 I thought this explained this issue.

    > regarding your questio. `TSO packet length` is it `dlen`? let's take an example in case `dlen==100000` and `mss=1500` I would send 66 windows without PUSH and residue with PUSH (the last).

    That's my suggestion! May I prepare a PR for this?
2. O.K.
3. O.K. let's focus on 1 at first.

**Comment by jsmoon** [comment]
@hhaim I've prepared a PR for 1. So, I think now we can discuss 3.

There are two ways in my mind. One is pre-processing and another is post-processing.
The pre-processing can be done by SLRO, merging ACK-only packets also.
The post-processing is somewhat complicated.
During RX packet handling in TCP_RX_FLUSH, the `tcp_output()` event should be suspended.
After the RX packet handling in a loop has been done, the suspended `tcp_output()` events can be performed.

@hhaim which one do you prefer?

**Comment by hhaim** [comment]
@jsmoon the PR looks good. 
Regarding the tcp_output(). We already have something like this in EMU layer. defer the ACK (only the biggest) would work here too

**Comment by jsmoon** [comment]
> Regarding the tcp_output(). We already have something like this in EMU layer.

I could not find the source code in trex-emu repository. Could you please let me know where it is?

> defer the ACK (only the biggest) would work here too

I've prepared a PR, the deferring ACK in software LRO implementation, #972.
@hhaim, please give me your thought about this change.
For your information, with this change, the queue_full case happens more frequently. I think this change increases TX's burstiness.

**Comment by hhaim** [comment]
@jsmoon sorry for the late response, I don't see how this helps having a bigger TSO response. 
merging the LRO **before** the TCP stack is better. 

Do you have an example cap file that shows better TSO usage?

**Comment by jsmoon** [comment]
> I don't see how this helps having a bigger TSO response.

@hhaim, this issue happens with small ACKs during the send window size is full. A small ACK will make a small room in the send window, and at that time, `tcp_output()` will send a small packet without TSO. Therefore, merging small ACKs to bigger one will make the room larger and `tcp_output()` can send a large packet with TSO. 

> merging the LRO before the TCP stack is better.

#972 is merging ACK (by the biggest) in the software LRO before the TCP stack. 

> Do you have an example cap file that shows better TSO usage?

I think you can get the pcap file by running `ftp_multiple_flows_sim.py` profile with the loopback environment.
Since #970 removes PUSH flag, to see the effect, you need to force PUSH flag by adding `tcp.no_delay = 2`.

**Comment by hhaim** [comment]
@jsmoon you are describing Silly Window Syndrome (SWS) which should be solved by the transmitter and not the receive side (wait for a bigger window). Even in my old freeBSD there was a solution for that

**Comment by jsmoon** [comment]
@hhaim I'm not sure what should I do next. Could you tell me more directly?
Should I abandon my change #972 and make a new solution by the transmitter?

**Comment by hhaim** [comment]
@jsmoon I don't want to invent a new TCP modes. At least we need to understand the issue. One of the reasons to move to a new TCP stack is to get new features and to simulate the stack more accurately. Adding this mode is diverting from the standard. SWS is solved by having delay ACK + ack every second packet by the transmitter. 

So my suggestion is first to understand if we have an issue with the current TCP implementation.

so I would return to https://github.com/cisco-system-traffic-generator/trex-core/blob/master/scripts/astf/http_eflow2.py
profile and understands how TSO works using this example (using the counters) when ACK is configured to be 

```
tcp.no_delay_counter  = 3000
tcp.no_delay =0 
tcp.delay_ack_msec = 20 
```

**Comment by jsmoon** [comment]
@hhaim `the current TCP implementation` means including PR #972? or not?

**Comment by jsmoon** [comment]
@hhaim before we are going further, let's check the following solution.

> SWS is solved by having delay ACK + ack every second packet by the transmitter.

"delay ACK" is not applied when the PUSH flag is marked. The original FreeBSD implementation does not have this behavior. As I understand it correctly, the PUSH flag is to use deliver data to the receiver application without buffering. But the old 4.4BSD implementation in T-Rex has an immediate ACK reply for the PUSH flagged packet, and it breaks the "delay ACK".

"ack every second packet" can be done by the current tunable as you mentioned before at https://github.com/cisco-system-traffic-generator/trex-core/issues/962#issuecomment-1396709517.
I think the tunable is `tcp.no_delay_counter`. But its behavior is slightly different from "ack every second packet". Since it counts received data bytes not the number of packets, ACK is replied by one single packet which is merged by LRO.

**Comment by hhaim** [comment]
@jsmoon 

1. "immediate ACK reply for the PUSH flagged packet" -- this is OK, this is the right behavior. 
2. setting the last window fragment with PUSH flag (not every packet), I think this is ok too, because trex known when this is the last segment of the window and this make the server side responds faster to finsh the flow.
3. "ack every second packet" => you are correct, you made somthing similar enogth for the RFC (bytes instead of packets) but you can tune it to be 2 by having the size of 2 MSS. LRO is an internal things so if you merge you still need to ACK every original second packet, so counting bytes is OK

I still does not understand the problem that you are trying to solve. I would take a eflow2.py profile and look into the TSO behaviour in this case

**Comment by jsmoon** [comment]
> "immediate ACK reply for the PUSH flagged packet" -- this is OK, this is the right behavior.

But I could not find the behavior in the original FreeBSD source code. I don't understand why FreeBSD does not have this behavior.

> I still does not understand the problem that you are trying to solve.

Actually, the original issue I had met was already solved by #970.
#972 is just an additional feature to solve another possible case like force PUSH is used with TSO.

**Comment by jsmoon** [comment]
@hhaim I'll close this issue because it is solved by #970.
Other derived issues need to be discussed in separate threads.
